### PR TITLE
CORE-7628 Add Support for TLS with IP Addresses

### PR DESCRIPTION
**Pull request by williamvigorr3**
If an ip address is specified as the host of the connectionURL in the members registration context (e.g. "corda.endpoints.0.connectionURL": "http://192.168.1.1:8080"), then the gateway can now connect to this member using an IP address.

To make this work: If an IP address is used when the gateway client to connects to the server then use an SNI of the IP address and a suffix of `.cluster.corda.net` e.g. `192.168.1.1.cluster.corda.net`. On the gateway server side if we detect the `.cluster.corda.net` suffix then we match a certificate with a Subject Alternative Name (SAN) with the IP in it. The server will only match certificates where the SAN contains an IP address not a DNS name. 

Also changed the RPC operation which generates a certificate signing request, so if an IP address is specified as a SAN  it is added as IP address not a DNS name.

Manual Testing
====

In addition to the integration test I added under the GatewayTests. I tested this manually using the following:
Deploy 3 clusters `williamvigor-cluster-a`, `williamvigor-cluster-b` and `williamvigor-mgm` (with 1 Gateway) into K8s. Find the IP addresses of the gateway pod in each cluster:
```
william.vigor@LDNM-WHD6VPFHQC scripts % kubectl describe pod corda-p2p-gateway-worker-77c85f8b8-px4wf -n williamvigor-mgm | grep IP
IP:           192.168.125.166
IPs:
  IP:           192.168.125.166
william.vigor@LDNM-WHD6VPFHQC scripts % kubectl describe pod corda-p2p-gateway-worker-77c85f8b8-vz6rg -n williamvigor-cluster-b | grep IP
IP:           192.168.108.139
IPs:
  IP:           192.168.108.139
william.vigor@LDNM-WHD6VPFHQC scripts % kubectl describe pod corda-p2p-gateway-worker-77c85f8b8-5xwv8 -n williamvigor-cluster-a | grep IP
IP:           192.168.104.68
IPs:
  IP:           192.168.104.68
```
During the MGM/Member onboarding steps: 

Replace `corda-p2p-gateway-worker.williamvigor-cluster-a` with `192.168.104.68`, `corda-p2p-gateway-worker.williamvigor-cluster-b` with `192.168.108.139`, corda-p2p-gateway-worker.williamvigor-mgm with `192.168.125.166`.

Replace `https://corda-p2p-gateway-worker.williamvigor-cluster-a:8080` with `https://192.168.104.68:8080`, `https://corda-p2p-gateway-worker.williamvigor-cluster-b` with `https://192.168.108.139:8080` and  `https://corda-p2p-gateway-worker.williamvigor-mgm:8080` `https://192.168.125.166:8080`, in the registration context for each cluster.

I checked 100 messages were successfully delivered using the app simulator.

### CODE REVIEW

**Review by wzur-r3 — DISMISSED** [review]
file `components/membership/membership-http-rpc-impl/build.gradle` LGTM

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt [hunk 1]
```diff
@@ -63,11 +71,34 @@ class HostnameMatcher(private val keyStore: KeyStore) : SNIMatcher(0) {
...
 
+    private fun matchIp(ip: String, certificate: X509Certificate): Boolean {
+        val names = certificate.subjectAlternativeNames
+        if (names.isNullOrEmpty()) {
+            logger.debug("No subject alternative names found in the certificate")
+            return false
+        }
+        names.forEach {
+            if (ALTNAME_IP == it[0]) {
+                val ipAddress = it[1] as String
```
Comment (yift-r3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt]: `it[1] as? String` just to be safe...

File: testing/p2p/certificates/src/main/resources/README.md [hunk 2]
```diff
@@ -22,6 +22,10 @@ OCSP responder.
     - status: OK
     - CN: Dale
     - Alternates: www.dale.net, dale.net, 127.0.0.1, https://127.0.0.1:10004
+- sslkeystore_127.0.0.1.jks
```
Comment (yift-r3) [inline testing/p2p/certificates/src/main/resources/README.md:25]: We can use a `CertificateAuthority` to create a certificate (that way we won't need to worry about expiry)

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt [hunk 3]
```diff
@@ -63,11 +71,34 @@ class HostnameMatcher(private val keyStore: KeyStore) : SNIMatcher(0) {
         return correctSize && correctSuffix && validHashedLegalName
     }
 
+    private fun isIpSni(serverName: String): Boolean {
+        val correctSuffix = serverName.endsWith(SniCalculator.IP_SNI_SUFFIX)
```
Comment (yift-r3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt]: Can we leave early if it's false?

Maybe do the `removeSuffix` and return false if it didn't change?

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt [hunk 4]
```diff
@@ -42,6 +45,11 @@ class HostnameMatcher(private val keyStore: KeyStore) : SNIMatcher(0) {
                 if (serverNameString == c4SniValue) {
                     return matched(alias)
                 }
+            } else if (isIpSni(serverNameString)) {
+                val ipAddress = serverNameString.removeSuffix(SniCalculator.IP_SNI_SUFFIX)
```
Comment (yift-r3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt:49]: if the `isIpSni` will return a `String?` after the `removeSuffix`, we will not have to do it again...

File: components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcherTest.kt [hunk 5]
```diff
@@ -30,6 +30,16 @@ class HostnameMatcherTest {
...
     }
 
+    @Test
+    fun `C5 SNI match with ip address`() {
+        val ipAddress = "127.0.0.1"
+        val keyStore: KeyStore = KeyStore.getInstance("JKS").also {
+            it.load(Certificates.ipKeyStore.openStream(), "password".toCharArray())
+        }
+        val matcher = HostnameMatcher(keyStore)
+        assertTrue(matcher.matches(SNIHostName(ipAddress + SniCalculator.IP_SNI_SUFFIX)))
```
Comment (yift-r3) [inline components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcherTest.kt]: `assertThat().isTrue`?

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt (hunk 4 above)
Comment (williamvigorr3) [reply components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/HostnameMatcher.kt:49]: I refactored this so we only do `removeSuffix` once.

File: testing/p2p/certificates/src/main/resources/README.md (hunk 2 above)
Comment (williamvigorr3) [reply testing/p2p/certificates/src/main/resources/README.md:25]: If we do that we need a different trust root. So I left is as is. We will tidy this stuff up as a part of https://r3-cev.atlassian.net/browse/CORE-5879.

**Review by yift-r3 — DISMISSED** [review]
LGTM - just need to solve the merge conflict

**Review by yift-r3 — APPROVED** [review]

**Review by mcgovc — APPROVED** [review]
LGTM - build changes

**Pull request merged**